analysis_driver/reader/demultiplexing_parsers.py

```diff
@@ -2,139 +2,146 @@
-                continue
+def parse_json_stats(json_data, run_id):
+    """Creates an array of tuples of run elements, unknown run elements and associated metadata."""
+    all_run_elements = []
+    unknown_run_elements = []
+    adapter_trimmed_by_id = {}
+
+    # Parsing conversion results, including demultiplexed results and unknown barcodes
+    for lane in json_data['ConversionResults']:
+        all_run_elements, adapter_trimmed_by_id = \
+            _parse_conversion_results_from_json_stats(run_id=run_id, lane=lane,
+                                                      all_run_elements=all_run_elements,
+                                                      adapter_trimmed_by_id=adapter_trimmed_by_id)
+
+        """ Including the values for unknown barcodes, only for multiplexed runs"""
+        if 'Undetermined' in lane:
+            all_run_elements, adapter_trimmed_by_id = \
+                _parse_undetermined_run_elements_from_json_stats(lane=lane,
+                                                                 run_id=run_id,
+                                                                 adapter_trimmed_by_id=adapter_trimmed_by_id,
+                                                                 all_run_elements=all_run_elements)
+
+    # parsing the top 10 unknown barcodes into their own array
+    unknown_run_elements = _parse_top_unknown_run_elements_from_json_stats(json_data=json_data)
+
+    return all_run_elements, unknown_run_elements, adapter_trimmed_by_id
+
+
+def _parse_conversion_results_from_json_stats(run_id, lane, all_run_elements, adapter_trimmed_by_id):
+    for sample in lane['DemuxResults']:
+        """ If the run is not multiplexed, the index_sequence will not be present in the JSON file, and the cluster
+        count needs to be retrieved from a different variable in the JSON file. The index_sequence variable will be 
+        prepopulated as None. """
+        r1, r2 = sample['ReadMetrics'][0], sample['ReadMetrics'][1]
+        index_sequence = None
+        cluster_count_raw = lane['TotalClustersRaw']
+        cluster_count_pf = lane['TotalClustersPF']
+        if 'IndexMetrics' in sample:
+            index_sequence = sample['IndexMetrics'][0]['IndexSequence']
+            cluster_count_raw = sample['NumberReads']
+            cluster_count_pf = sample['NumberReads']  # cluster count is equal to cluster count pf in multiplexed runs
+
+        # obtaining number of trimmed bases and bases r1 and r2 q30, confirming the read number first
+        nb_bases_r1_q30, nb_bases_r2_q30 = 0, 0
+        if r1['ReadNumber'] == 1 and r2['ReadNumber'] == 2:
+            nb_bases_r1_q30 = r1['YieldQ30']
+            nb_bases_r2_q30 = r2['YieldQ30']
+            trimmed_bases_r1 = r1['TrimmedBases']
+            trimmed_bases_r2 = r2['TrimmedBases']
+        elif r1['ReadNumber'] == 2 and r2['ReadNumber'] == 1:
+            # this is not expected
+            raise NotImplementedError()
```

Comment:
NotImplementedError is [intended for defining abstract methods](https://docs.python.org/3/library/exceptions.html?highlight=notimplementederror#NotImplementedError) - I'd raise something like `SequencingRunError`. Not too important right now, though - can refactor later

Reply:
@mwhamgenomics I agree, that's why I placed it in a helper function so at least it's only found once in the source code.

Reply:
@mwhamgenomics Technically this isn't an error. There would be nothing wrong if the first ReadNumber is actually the R2 value, and the second ReadNumber is the R1 value. I was not sure whether this could ever happen, so I thought it would be best to handle the case and raise a NotImplementedError() if we ever encounter that, so we can implement it.